lib/RestructureCFG/CycleEquivalenceAnalysis.cpp

```diff
@@ -0,0 +1,468 @@
+                                    << Edge.second->getName() << "),"
+                                    << std::to_string(Size) << ")");
+  }
+}
+
+template<class GraphT, class GT>
+bool CycleEquivalenceAnalysis<GraphT, GT>::isBackedge(EdgeDescriptor &ED) {
+
+  size_t SourceDFSNum = NodeDFSNumMap.at(ED.first);
+  size_t TargetDFSNum = NodeDFSNumMap.at(ED.second);
+
+  return TargetDFSNum < SourceDFSNum;
+}
+
+template<class GraphT, class GT>
+void CycleEquivalenceAnalysis<GraphT, GT>::computeDFSAndSpanningTree(BlockGraph
+                                                                       &Graph) {
+  using ExtType = llvm::df_iterator_default_set<BlockNode *>;
+  using UGT = llvm::GraphTraits<llvm::Undirected<BlockNode *>>;
+  using udf_iterator = llvm::df_iterator<BlockNode *, ExtType, false, UGT>;
+
+  BlockNode *EntryNode = Graph.getEntryNode();
+  auto Begin = udf_iterator::begin(EntryNode);
+  auto End = udf_iterator::end(EntryNode);
+
+  size_t DFSNum = 0;
+  for (; Begin != End; Begin++) {
+
+    // We assign the DFS numbering
+
+    BlockNode *CurrentNode = *Begin;
+
+    NodeDFSNumMap[CurrentNode] = DFSNum;
+    DFSNumNodeMap[DFSNum] = CurrentNode;
+    DFSNum++;
+
+    // Each time we reach a new node during the `DFS` visit, we search for the
+    // edge that brougth us here, and that we therefore need to insert into the
+    // spanning tree.
+    // Retrieve the last-but-one element on the `VisitStack`, which is the
+    // candidate as the source node of the edge we are interested into adding to
+    // the spanning tree.
+    auto VisitStackSize = Begin.getPathLength();
+
+    // When we have a single node on the `VisitStack`, we cannot assign an edge
+    // to the spanning tree
+    if (VisitStackSize < 2)
+      continue;
+
+    BlockNode *SourceNode = Begin.getPath(VisitStackSize - 2);
+
+    // Search for the edge which connects `Source` and `CurrentNode`
+    for (auto &[Neighbor, Index] :
+         llvm::children_edges<llvm::Undirected<BlockNode *>>(SourceNode)) {
+
+      // Check if the successor is equal to `CurrentNode`, and in case add it to
+      // the spanning tree
+      if (Neighbor == CurrentNode) {
+        SpanningTreeEdges.insert(Index);
+        break;
+      }
+    }
+  }
+
+  // Printout the spanning tree
+  revng_log(CycleEquivalenceAnalysisLogger, "");
+  revng_log(CycleEquivalenceAnalysisLogger,
+            "Spanning tree is composed by edges:");
+  for (auto &Index : SpanningTreeEdges) {
+    revng_log(CycleEquivalenceAnalysisLogger,
+              std::to_string(Index.getValue()) << "\n");
+  }
+}
+
+/// Helper function to compute the capping backedges needed values
+template<class GraphT, class GT>
+void CycleEquivalenceAnalysis<GraphT, GT>::computeCappingValues(BlockNode *BB,
+                                                                size_t &HI0,
+                                                                size_t &HI2) {
```

Comment:
Make this return a `std::pair<size_t, size_t>` or even better, a plain `struct` with a meaningful name, and 2 `size_t` fields.

Reply:
This method has been dropped entirely, to I think this is updated, the two parameters are now simply local variables in this https://github.com/revng/revng-c/blob/323b04da3a0979143e3207968b19e7142893109e/lib/RestructureCFG/CycleEquivalenceAnalysis.cpp#L162C44-L162C66 method

Reply:
:+1: